subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/scheduler/scheduler.go

```diff
@@ -671,10 +685,66 @@ func (s *HivedScheduler) preemptRoutine(args ei.ExtenderPreemptionArgs) *ei.Exte
 	return &ei.ExtenderPreemptionResult{}
 }
 
-func (s *HivedScheduler) getAffinityGroups() si.AffinityGroupList {
-	return s.schedulerAlgorithm.GetAffinityGroups()
+func (s *HivedScheduler) getAllAffinityGroups() si.AffinityGroupList {
+	return s.schedulerAlgorithm.GetAllAffinityGroups()
 }
 
 func (s *HivedScheduler) getAffinityGroup(name string) si.AffinityGroup {
 	return s.schedulerAlgorithm.GetAffinityGroup(name)
 }
+
+func (s *HivedScheduler) getClusterStatus() si.ClusterStatus {
+	return s.schedulerAlgorithm.GetClusterStatus()
+}
+
+func (s *HivedScheduler) watchNodes() {
+	for {
+		s.watchHealthyNodes()
+		s.watchBadNodes()
+		time.Sleep(watchNodesInterval * time.Second)
+	}
+}
+
+// watchHealthyNodes checks the healthiness of nodes. It it found a bad node,
+// it will call the scheduler algorithm to mark it as node.
+func (s *HivedScheduler) watchHealthyNodes() {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	var badNodeIndices []int32
+	for i, n := range s.healthyNodes {
+		if _, err := s.nodeLister.Get(n); err != nil {
+			s.schedulerAlgorithm.SetBadNode(n)
+			badNodeIndices = append(badNodeIndices, int32(i))
+		}
+	}
+	oriHealthyNodeNum := len(s.healthyNodes)
+	for i, bni := range badNodeIndices {
+		s.badNodes = append(s.badNodes, s.healthyNodes[bni])
+		s.healthyNodes[bni] = s.healthyNodes[oriHealthyNodeNum-1-i]
+		s.healthyNodes[oriHealthyNodeNum-1-i] = ""
+	}
+	s.healthyNodes = s.healthyNodes[:oriHealthyNodeNum-len(badNodeIndices)]
+}
+
+// watchBadNodes checks the healthiness of bad nodes. If a bad node is found to be healthy again,
+// we will call the scheduler algorithm to mark it as healthy.
+func (s *HivedScheduler) watchBadNodes() {
+	s.schedulerLock.Lock()
+	defer s.schedulerLock.Unlock()
+
+	var healthyNodeIndices []int32
+	for i, n := range s.badNodes {
+		if _, err := s.nodeLister.Get(n); err == nil {
+			s.schedulerAlgorithm.SetHealthyNode(n)
+			healthyNodeIndices = append(healthyNodeIndices, int32(i))
+		}
+	}
+	oriBadNodeNum := len(s.badNodes)
+	for i, hni := range healthyNodeIndices {
+		s.healthyNodes = append(s.healthyNodes, s.badNodes[hni])
+		s.badNodes[hni] = s.badNodes[oriBadNodeNum-1-i]
+		s.badNodes[oriBadNodeNum-1-i] = ""
```

Comment:
>s.badNodes[oriBadNodeNum-1-i] = "" [](start = 2, length = 34)

Why need s.badNodes[oriBadNodeNum-1-i] = ""

Reply:
Solved as this logic has gone